File: cp-trials/src/test/java/in/shabhushan/cp_trials/dynamic_programming/CoinChangeTest.java
package in.shabhushan.cp_trials.dynamic_programming;

import org.junit.Test;

import static in.shabhushan.cp_trials.dynamic_programming.CoinChange.*;
import static org.junit.Assert.assertEquals;

public class CoinChangeTest {
  @Test
  public void testCoinChangeRecursiveWithMemoization() {
    assertEquals(2, minimumCoinChange(new int[]{1, 3, 4}, 6));
    assertEquals(2, minimumCoinChange(new int[]{1, 3, 4}, 7));
    assertEquals(3, minimumCoinChange(new int[]{1, 2, 5}, 11));
    assertEquals(-1, minimumCoinChange(new int[]{2}, 3));
  }

  @Test
  public void testCoinChangecoinChangeIterative() {
    assertEquals(2, coinChangeIterative(new int[]{1, 3, 4}, 6));
    assertEquals(2, coinChangeIterative(new int[]{1, 3, 4}, 7));
    assertEquals(3, coinChangeIterative(new int[]{1, 2, 5}, 11));
    assertEquals(-1, coinChangeIterative(new int[]{2}, 3));
  }

  @Test
  public void testCoinChangeWithSequence() {
    CoinChangeSolution solution = coinChangeIterativeWithCounting(new int[]{1, 3, 4}, 6);
    assertEquals(2, solution.solution);
    int target = 6;

    while (target > 0) {
      System.out.println(solution.sequence[target]);
      target -= solution.sequence[target];
    }
    CoinChangeSolution solution2 = coinChangeIterativeWithCounting(new int[]{186, 419, 83, 408}, 6249);
    assertEquals(20, solution2.solution);
    int target2 = 6249;

    while (target2 > 0) {
      System.out.println(solution2.sequence[target2]);
      target2 -= solution2.sequence[target2];
    }
  }

  @Test
  public void testCoinChangeTotalWays() {
    assertEquals(6, totalWaysCoinChange(new int[]{1, 3, 4}, 5));
  }

  @Test
  public void testCoinChangeMinimumCoins() {
    assertEquals(3, coinChangeMinimumCoins(new int[]{1, 2, 5}, 11));
    assertEquals(-1, coinChangeMinimumCoins(new int[]{2}, 3));
    //assertEquals(20, coinChangeMinimumCoins(new int[]{186, 419, 83, 408}, 6249));
  }

  @Test
  public void testCoinChangeMinimumCoinsDP() {
    assertEquals(3, coinChangeMinimumCoinsDP(new int[]{1, 2, 5}, 11));
    assertEquals(-1, coinChangeMinimumCoinsDP(new int[]{2}, 3));
    assertEquals(20, coinChangeMinimumCoinsDP(new int[]{186, 419, 83, 408}, 6249));
  }

  @Test
  public void testChange() {
    assertEquals(4, change(new int[]{1, 2, 5}, 5));
    assertEquals(0, change(new int[]{2}, 3));
    assertEquals(1, change(new int[]{10}, 10));
  }

  @Test
  public void testChangeRecursive() {
    assertEquals(4, changeRecursive(new int[]{1, 2, 5}, 5));
    assertEquals(0, changeRecursive(new int[]{2}, 3));
    assertEquals(1, changeRecursive(new int[]{10}, 10));
  }

  @Test
  public void testNumSquares() {
    assertEquals(2, numSquares(5));
    assertEquals(3, numSquares(11));
  }

}

File: cp-trials/src/test/kotlin/in/shabhushan/cp_trials/dsbook/methods/backtracking/CoinChangeTest.kt
package `in`.shabhushan.cp_trials.dsbook.methods.backtracking

import `in`.shabhushan.cp_trials.dsbook.methods.backtracking.CoinChange.coinChangeMinimumCoins
import `in`.shabhushan.cp_trials.dsbook.methods.backtracking.CoinChange.coinChangePossible
import `in`.shabhushan.cp_trials.dsbook.methods.backtracking.CoinChange.coinChangeTotalCanonicallySimilarWays
import `in`.shabhushan.cp_trials.dsbook.methods.backtracking.CoinChange.coinChangeTotalWays
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class CoinChangeTest {
  @Test
  fun testCoinChange_Possible() {
    assertTrue(coinChangePossible(intArrayOf(5, 2, 2), 9))
    assertTrue(coinChangePossible(intArrayOf(5, 2, 2), 7))
    assertFalse(coinChangePossible(intArrayOf(5, 2, 2), 3))
  }

  @Test
  fun testCoinChange_MinimumCoins() {
    assertEquals(2, coinChangeMinimumCoins(intArrayOf(1, 3, 4), 6))
    assertEquals(1, coinChangeMinimumCoins(intArrayOf(1, 2, 5), 5))
  }

  @Test
  fun testCoinChange_TotalCanonicallySimilarWays() {
    //assertEquals(4, coinChangeTotalCanonicallySimilarWays(intArrayOf(1, 3, 4), 6))
    assertEquals(4, coinChangeTotalCanonicallySimilarWays(intArrayOf(1, 2, 5), 5))
    //assertEquals(0, coinChangeTotalCanonicallySimilarWays(intArrayOf(2, 5), 3))
  }

  @Test
  fun testCoinChange_TotalWays() {
    //assertEquals(9, coinChangeTotalWays(intArrayOf(1, 3, 4), 6))
    assertEquals(9, coinChangeTotalWays(intArrayOf(1, 2, 5), 5))
    //assertEquals(0, coinChangeTotalWays(intArrayOf(2, 5), 3))
  }
}
